assistant/assistant.go

```diff
@@ -0,0 +1,186 @@
+package assistant
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/influxdata/telegraf/agent"
+	"github.com/influxdata/telegraf/internal"
+)
+
+/*
+Assistant is a client to facilitate communications between Agent and Cloud.
+*/
+type Assistant struct {
+	config     *AssistantConfig // stores plugins
+	connection *websocket.Conn  // Active websocket connection
+	ctx        context.Context  // go's context
+	done       chan bool        // Channel used to stop server listener
+	agent      *agent.Agent     // Pointer to agent to issue commands
+}
+
+/*
+AssistantConfig allows us to configure where to connect to and other params
+for the agent.
+*/
+type AssistantConfig struct {
+	Host          string
+	Path          string
+	RetryInterval int
+}
+
+func (astConfig *AssistantConfig) fillDefaults() {
+	if astConfig.Host == "" {
+		astConfig.Host = "localhost:8080"
+	}
+	if astConfig.Path == "" {
+		astConfig.Path = "/echo"
+	}
+	if astConfig.RetryInterval == 0 {
+		astConfig.RetryInterval = 15
+	}
+}
+
+// NewAssistant returns an Assistant for the given Config.
+func NewAssistant(ctx context.Context, config *AssistantConfig, agent *agent.Agent) (*Assistant, error) {
+	config.fillDefaults()
+	var addr = flag.String("addr", config.Host, "http service address")
+	u := url.URL{Scheme: "ws", Host: *addr, Path: config.Path}
+
+	header := http.Header{}
+
+	if v, exists := os.LookupEnv("INFLUX_TOKEN"); exists {
+		header.Add("Authorization", "Token "+v)
+	} else {
+		return nil, fmt.Errorf("Influx authorization token not found, please set in env")
+	}
+
+	// creates a new websocket connection
+	log.Printf("D! [assistant] Attempting connection to [%s]", config.Host)
+	ws, _, err := websocket.DefaultDialer.Dial(u.String(), header)
```

Comment:
What's the second argument being ignored?

Reply:
the 2nd argument returns the `http.response` from attempting to create the connection. According to the documentation you can use it to get things like protocols and cookies which i don't think we need